File: train_notify/core/context.py
from __future__ import annotations

import os
import shlex
import socket
import subprocess
from datetime import datetime, timezone
from pathlib import Path
from typing import Iterable, Optional

from .redact import redact_text


def now_iso() -> str:
    return datetime.now(timezone.utc).isoformat()


def _run_git(args: list[str], cwd: Path) -> Optional[str]:
    try:
        proc = subprocess.run(
            ["git", *args],
            cwd=str(cwd),
            check=False,
            capture_output=True,
            text=True,
        )
    except Exception:
        return None
    if proc.returncode != 0:
        return None
    value = proc.stdout.strip()
    return value or None


def detect_project_name(cwd: str | Path) -> str:
    path = Path(cwd).resolve()
    top_level = _run_git(["rev-parse", "--show-toplevel"], cwd=path)
    if top_level:
        return Path(top_level).name
    return path.name


def detect_git_branch(cwd: str | Path) -> Optional[str]:
    return _run_git(["rev-parse", "--abbrev-ref", "HEAD"], cwd=Path(cwd))


def detect_git_commit(cwd: str | Path) -> Optional[str]:
    return _run_git(["rev-parse", "--short", "HEAD"], cwd=Path(cwd))


def infer_job_name(cmd: Iterable[str]) -> str:
    parts = list(cmd)
    if not parts:
        return "unknown-job"

    first = Path(parts[0]).name
    if first == "uv" and len(parts) >= 3 and parts[1] == "run":
        return infer_job_name(parts[2:])

    if first in {"python", "python3", "python3.10", "python3.11", "uv"}:
        for token in parts[1:]:
            if token.startswith("-"):
                continue
            return Path(token).stem
    if first == "conda":
        if "run" in parts:
            run_index = parts.index("run")
            rest = parts[run_index + 1 :]
            value_options = {"-n", "--name", "-p", "--prefix"}
            idx = 0
            while idx < len(rest) and rest[idx].startswith("-"):
                opt = rest[idx]
                idx += 2 if opt in value_options and idx + 1 < len(rest) else 1
            if idx < len(rest):
                return infer_job_name(rest[idx:])
            return "conda-task"
        for idx, token in enumerate(parts[1:], start=1):
            if token.startswith("-"):
                continue
            return Path(token).stem
        return "conda-task"
    return Path(parts[0]).stem


def format_cmd(cmd: Iterable[str], redact_patterns: list[str]) -> str:
    raw = shlex.join(list(cmd))
    return redact_text(raw, redact_patterns)


def build_run_context(
    run_id: str,
    job_name: str,
    cmd: Iterable[str],
    cwd: str | Path,
    log_path: Optional[str],
    redact_patterns: list[str],
    tmux_session: Optional[str] = None,
    pid: Optional[int] = None,
) -> dict:
    cwd_path = Path(cwd).resolve()
    return {
        "run_id": run_id,
        "project": detect_project_name(cwd_path),
        "job_name": job_name,
        "host": socket.gethostname(),
        "cwd": str(cwd_path),
        "git_branch": detect_git_branch(cwd_path),
        "git_commit": detect_git_commit(cwd_path),
        "cmd": format_cmd(cmd, redact_patterns),
        "log_path": str(Path(log_path).resolve()) if log_path else None,
        "tmux_session": tmux_session or os.getenv("TRAIN_NOTIFY_TMUX_SESSION"),
        "pid": pid,
        "start_time": now_iso(),
    }

File: internal/context/context.go
package context

import (
	"bytes"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"slices"
	"strings"
	"time"
)

type RunContext struct {
	RunID       string
	Project     string
	JobName     string
	Host        string
	CWD         string
	GitBranch   string
	GitCommit   string
	Cmd         string
	LogPath     string
	TmuxSession string
	PID         int
	StartTime   string
}

var utcPlus8 = time.FixedZone("UTC+8", 8*3600)

func nowISO() string {
	return time.Now().In(utcPlus8).Format(time.RFC3339)
}

func runGit(cwd string, args ...string) string {
	cmd := exec.Command("git", args...)
	cmd.Dir = cwd
	var out bytes.Buffer
	cmd.Stdout = &out
	if err := cmd.Run(); err != nil {
		return ""
	}
	return strings.TrimSpace(out.String())
}

func DetectProjectName(cwd string) string {
	abs, err := filepath.Abs(cwd)
	if err != nil {
		abs = cwd
	}
	top := runGit(abs, "rev-parse", "--show-toplevel")
	if top != "" {
		return filepath.Base(top)
	}
	return filepath.Base(abs)
}

func DetectGitBranch(cwd string) string {
	return runGit(cwd, "rev-parse", "--abbrev-ref", "HEAD")
}

func DetectGitCommit(cwd string) string {
	return runGit(cwd, "rev-parse", "--short", "HEAD")
}

func InferJobName(parts []string) string {
	if len(parts) == 0 {
		return "unknown-job"
	}
	first := filepath.Base(parts[0])
	if first == "uv" && len(parts) >= 3 && parts[1] == "run" {
		return InferJobName(parts[2:])
	}
	if first == "python" || first == "python3" || strings.HasPrefix(first, "python3.") || first == "uv" {
		for _, token := range parts[1:] {
			if strings.HasPrefix(token, "-") {
				continue
			}
			return strings.TrimSuffix(filepath.Base(token), filepath.Ext(token))
		}
	}
	if first == "conda" {
		idx := slices.Index(parts, "run")
		if idx >= 0 {
			rest := parts[idx+1:]
			valueOptions := map[string]bool{"-n": true, "--name": true, "-p": true, "--prefix": true}
			i := 0
			for i < len(rest) && strings.HasPrefix(rest[i], "-") {
				if valueOptions[rest[i]] && i+1 < len(rest) {
					i += 2
				} else {
					i++
				}
			}
			if i < len(rest) {
				return InferJobName(rest[i:])
			}
			return "conda-task"
		}
		for _, token := range parts[1:] {
			if strings.HasPrefix(token, "-") {
				continue
			}
			return strings.TrimSuffix(filepath.Base(token), filepath.Ext(token))
		}
		return "conda-task"
	}
	return strings.TrimSuffix(filepath.Base(parts[0]), filepath.Ext(parts[0]))
}

func JoinCommand(parts []string) string {
	if len(parts) == 0 {
		return ""
	}
	quoted := make([]string, 0, len(parts))
	for _, p := range parts {
		quoted = append(quoted, shellQuote(p))
	}
	return strings.Join(quoted, " ")
}

func shellQuote(s string) string {
	if s == "" {
		return "''"
	}
	if strings.IndexFunc(s, func(r rune) bool {
		return r == ' ' || r == '\t' || r == '\n' || r == '\\' || r == '"' || r == '\'' || r == '$' || r == '`' || r == '!' || r == '|' || r == '&' || r == ';' || r == '<' || r == '>' || r == '(' || r == ')'
	}) == -1 {
		return s
	}
	return "'" + strings.ReplaceAll(s, "'", "'\"'\"'") + "'"
}

func BuildRunContext(runID, jobName string, cmd []string, cwd, logPath string, redactPatterns []string, tmuxSession string, pid int) (RunContext, error) {
	absCWD, err := filepath.Abs(cwd)
	if err != nil {
		return RunContext{}, fmt.Errorf("resolve cwd: %w", err)
	}
	host, _ := os.Hostname()
	lp := ""
	if logPath != "" {
		abslp, err := filepath.Abs(logPath)
		if err != nil {
			return RunContext{}, fmt.Errorf("resolve log path: %w", err)
		}
		lp = abslp
	}
	if tmuxSession == "" {
		tmuxSession = os.Getenv("TRAINPULSE_TMUX_SESSION")
	}
	return RunContext{
		RunID:       runID,
		Project:     DetectProjectName(absCWD),
		JobName:     jobName,
		Host:        host,
		CWD:         absCWD,
		GitBranch:   DetectGitBranch(absCWD),
		GitCommit:   DetectGitCommit(absCWD),
		Cmd:         RedactText(JoinCommand(cmd), redactPatterns),
		LogPath:     lp,
		TmuxSession: tmuxSession,
		PID:         pid,
		StartTime:   nowISO(),
	}, nil
}
